### DISCUSSION TITLE
android build 2.10.2 + TMS + TFS data leads to regular crashes

### DISCUSSION THREAD

**Discussion by shajder** [pelicanmapping/osgearth#2168] [Support Q&A] [answered]
Hello Dear Community

I am building an android app which suppose to show map from TMS (image + elevation from http) and TFS (labels stored in apk) data. Good news is that I can see the correct content in application but the problem is it crashes sooner or later. Crash is related to threading and after many launches I isolated two colliding threads:

thread1 : OSG cull traversal -> invalid Group::_children iterator -> crash (child was added/removed during traversal?)
thread2 : osgEarthFeatureModelPseudoLoader -> ElevationPool processing

This should work because it seems like thread2 does not modify scene graph until OSG update traversal trigger DatabasePager::addLoadedDataToSceneGraph which is correct moment. Still it crashes.

Additional info is that I am launching windows build with the same source code and data - it works without glitches obviously. 

Is there any chance someone faced similar problem ? I would appreciate any hint here. Thanks

cheers
Marcin

**Comment by shajder** [comment, marked as answer]
Hello again

Although this is probably not a former solution I can give some clue to save someone's time. 

After some laborious investigation I eliminated the crashes by commenting `_OSG_REFERENCED_USE_ATOMIC_OPERATIONS` flag from **include/Referenced** header. To build OSG/osgEarth for Android I am using Android Studio which uses Clang compiler - that's probably significant in this case.

I hope this could help.

Cheers
Marcin

**Reply by gwaldron** [reply]
A quick glance and it seems like OpenThreads/Atomic could probably use an update to use `std::atomic` across supported platforms.
